{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9144c779-0054-4129-bff2-1a7d4d49d1f3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: sanbomics in c:\\python312\\lib\\site-packages (0.1.0)\n",
      "Requirement already satisfied: pandas>=1.0 in c:\\python312\\lib\\site-packages (from sanbomics) (2.1.4)\n",
      "Requirement already satisfied: numpy>=1.17.0 in c:\\python312\\lib\\site-packages (from sanbomics) (1.26.3)\n",
      "Requirement already satisfied: matplotlib>=3.4 in c:\\python312\\lib\\site-packages (from sanbomics) (3.8.4)\n",
      "Requirement already satisfied: seaborn in c:\\python312\\lib\\site-packages (from sanbomics) (0.13.2)\n",
      "Requirement already satisfied: adjustText>=0.7.3 in c:\\python312\\lib\\site-packages (from sanbomics) (1.1.1)\n",
      "Requirement already satisfied: scipy>=1.9.1 in c:\\python312\\lib\\site-packages (from sanbomics) (1.12.0)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in c:\\python312\\lib\\site-packages (from matplotlib>=3.4->sanbomics) (1.2.0)\n",
      "Requirement already satisfied: cycler>=0.10 in c:\\python312\\lib\\site-packages (from matplotlib>=3.4->sanbomics) (0.12.1)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in c:\\python312\\lib\\site-packages (from matplotlib>=3.4->sanbomics) (4.49.0)\n",
      "Requirement already satisfied: kiwisolver>=1.3.1 in c:\\python312\\lib\\site-packages (from matplotlib>=3.4->sanbomics) (1.4.5)\n",
      "Requirement already satisfied: packaging>=20.0 in c:\\python312\\lib\\site-packages (from matplotlib>=3.4->sanbomics) (23.2)\n",
      "Requirement already satisfied: pillow>=8 in c:\\python312\\lib\\site-packages (from matplotlib>=3.4->sanbomics) (10.2.0)\n",
      "Requirement already satisfied: pyparsing>=2.3.1 in c:\\python312\\lib\\site-packages (from matplotlib>=3.4->sanbomics) (3.1.1)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in c:\\python312\\lib\\site-packages (from matplotlib>=3.4->sanbomics) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in c:\\python312\\lib\\site-packages (from pandas>=1.0->sanbomics) (2023.3.post1)\n",
      "Requirement already satisfied: tzdata>=2022.1 in c:\\python312\\lib\\site-packages (from pandas>=1.0->sanbomics) (2023.4)\n",
      "Requirement already satisfied: six>=1.5 in c:\\python312\\lib\\site-packages (from python-dateutil>=2.7->matplotlib>=3.4->sanbomics) (1.16.0)\n"
     ]
    }
   ],
   "source": [
    "!pip install sanbomics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "69f6bdb4-edbc-40d4-88cb-c930e317e19c",
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'sanbomics'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[2], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01msanbomics\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtools\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m id_map\n",
      "\u001b[1;31mModuleNotFoundError\u001b[0m: No module named 'sanbomics'"
     ]
    }
   ],
   "source": [
    "from sanbomics.tools import id_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd16cf0f-6592-4208-993a-29afa615a9c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from gseapy.plot import gseaplot\n",
    "import gseapy as gp\n",
    "import numpy as np\n",
    "from pydeseq2.dds import DeseqDataSet\n",
    "from pydeseq2.ds import DeseqStats\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d301ab92-a28e-4d05-b979-8281e58b0109",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts = pd.read_csv('Merged.csv')\n",
    "counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3383c2c-4a9d-45bd-9318-1a525484e645",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts = counts.dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f2a9673a-8ddb-4346-9722-6347660a85b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SRR11235153_Normal</th>\n",
       "      <th>SRR11235154_Normal</th>\n",
       "      <th>SRR11235155_Disease</th>\n",
       "      <th>SRR11235156_Disease</th>\n",
       "      <th>SRR11235157_Disease</th>\n",
       "      <th>SRR11235158_Disease</th>\n",
       "      <th>SRR11235159_Disease</th>\n",
       "      <th>SRR11235160_Disease</th>\n",
       "      <th>SRR19591720_Disease</th>\n",
       "      <th>SRR19591720_Disease.1</th>\n",
       "      <th>...</th>\n",
       "      <th>SRR12362564_Disease.1</th>\n",
       "      <th>SRR12362564_Disease.2</th>\n",
       "      <th>SRR23920508_1_Disease</th>\n",
       "      <th>SRR23920508_1_Disease.1</th>\n",
       "      <th>SRR23920508_1_Disease.2</th>\n",
       "      <th>SRR23920508_1_Disease.3</th>\n",
       "      <th>SRR23920508_1_Disease.4</th>\n",
       "      <th>SRR23920508_1_Disease.5</th>\n",
       "      <th>SRR23920508_1_Disease.6</th>\n",
       "      <th>SRR23920508_1_Disease.7</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>1352.0</td>\n",
       "      <td>1378.0</td>\n",
       "      <td>1382.0</td>\n",
       "      <td>1362.0</td>\n",
       "      <td>2499.0</td>\n",
       "      <td>2633.0</td>\n",
       "      <td>1721.0</td>\n",
       "      <td>1817.0</td>\n",
       "      <td>1567.0</td>\n",
       "      <td>1568.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2031</td>\n",
       "      <td>2032</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005</th>\n",
       "      <td>145.0</td>\n",
       "      <td>164.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>846.0</td>\n",
       "      <td>847.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2031</td>\n",
       "      <td>2032</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>1320.0</td>\n",
       "      <td>1307.0</td>\n",
       "      <td>1087.0</td>\n",
       "      <td>1206.0</td>\n",
       "      <td>1285.0</td>\n",
       "      <td>1312.0</td>\n",
       "      <td>1148.0</td>\n",
       "      <td>1174.0</td>\n",
       "      <td>5587.0</td>\n",
       "      <td>5588.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <td>183.0</td>\n",
       "      <td>190.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>168.0</td>\n",
       "      <td>195.0</td>\n",
       "      <td>179.0</td>\n",
       "      <td>175.0</td>\n",
       "      <td>188.0</td>\n",
       "      <td>3188.0</td>\n",
       "      <td>3189.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>81.0</td>\n",
       "      <td>84.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>84.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>1209.0</td>\n",
       "      <td>1210.0</td>\n",
       "      <td>...</td>\n",
       "      <td>6541</td>\n",
       "      <td>6542</td>\n",
       "      <td>10.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>17.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288721</th>\n",
       "      <td>52.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>83.0</td>\n",
       "      <td>108.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>63.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>421.0</td>\n",
       "      <td>422.0</td>\n",
       "      <td>...</td>\n",
       "      <td>24</td>\n",
       "      <td>25</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288722</th>\n",
       "      <td>253.0</td>\n",
       "      <td>212.0</td>\n",
       "      <td>198.0</td>\n",
       "      <td>227.0</td>\n",
       "      <td>450.0</td>\n",
       "      <td>500.0</td>\n",
       "      <td>493.0</td>\n",
       "      <td>459.0</td>\n",
       "      <td>120.0</td>\n",
       "      <td>121.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288723</th>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288724</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>33</td>\n",
       "      <td>34</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288725</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>200</td>\n",
       "      <td>201</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60664 rows × 61 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 SRR11235153_Normal  SRR11235154_Normal  SRR11235155_Disease  \\\n",
       "ENSG_ID                                                                        \n",
       "ENSG00000000003              1352.0              1378.0               1382.0   \n",
       "ENSG00000000005               145.0               164.0                  7.0   \n",
       "ENSG00000000419              1320.0              1307.0               1087.0   \n",
       "ENSG00000000457               183.0               190.0                200.0   \n",
       "ENSG00000000460                81.0                84.0                 43.0   \n",
       "...                             ...                 ...                  ...   \n",
       "ENSG00000288721                52.0                77.0                 54.0   \n",
       "ENSG00000288722               253.0               212.0                198.0   \n",
       "ENSG00000288723                 0.0                 3.0                  0.0   \n",
       "ENSG00000288724                 0.0                 0.0                  0.0   \n",
       "ENSG00000288725                 0.0                 0.0                  0.0   \n",
       "\n",
       "                 SRR11235156_Disease  SRR11235157_Disease  \\\n",
       "ENSG_ID                                                     \n",
       "ENSG00000000003               1362.0               2499.0   \n",
       "ENSG00000000005                 13.0                 52.0   \n",
       "ENSG00000000419               1206.0               1285.0   \n",
       "ENSG00000000457                168.0                195.0   \n",
       "ENSG00000000460                 59.0                 93.0   \n",
       "...                              ...                  ...   \n",
       "ENSG00000288721                 83.0                108.0   \n",
       "ENSG00000288722                227.0                450.0   \n",
       "ENSG00000288723                  2.0                  0.0   \n",
       "ENSG00000288724                  0.0                  0.0   \n",
       "ENSG00000288725                  0.0                  0.0   \n",
       "\n",
       "                 SRR11235158_Disease  SRR11235159_Disease  \\\n",
       "ENSG_ID                                                     \n",
       "ENSG00000000003               2633.0               1721.0   \n",
       "ENSG00000000005                 57.0                 77.0   \n",
       "ENSG00000000419               1312.0               1148.0   \n",
       "ENSG00000000457                179.0                175.0   \n",
       "ENSG00000000460                 84.0                 31.0   \n",
       "...                              ...                  ...   \n",
       "ENSG00000288721                 72.0                 63.0   \n",
       "ENSG00000288722                500.0                493.0   \n",
       "ENSG00000288723                  0.0                  0.0   \n",
       "ENSG00000288724                  0.0                  0.0   \n",
       "ENSG00000288725                  0.0                  0.0   \n",
       "\n",
       "                 SRR11235160_Disease  SRR19591720_Disease  \\\n",
       "ENSG_ID                                                     \n",
       "ENSG00000000003               1817.0               1567.0   \n",
       "ENSG00000000005                 95.0                846.0   \n",
       "ENSG00000000419               1174.0               5587.0   \n",
       "ENSG00000000457                188.0               3188.0   \n",
       "ENSG00000000460                 34.0               1209.0   \n",
       "...                              ...                  ...   \n",
       "ENSG00000288721                 65.0                421.0   \n",
       "ENSG00000288722                459.0                120.0   \n",
       "ENSG00000288723                  0.0                  0.0   \n",
       "ENSG00000288724                  0.0                  0.0   \n",
       "ENSG00000288725                  0.0                  0.0   \n",
       "\n",
       "                 SRR19591720_Disease.1  ...  SRR12362564_Disease.1  \\\n",
       "ENSG_ID                                 ...                          \n",
       "ENSG00000000003                 1568.0  ...                   2031   \n",
       "ENSG00000000005                  847.0  ...                   2031   \n",
       "ENSG00000000419                 5588.0  ...                      1   \n",
       "ENSG00000000457                 3189.0  ...                      1   \n",
       "ENSG00000000460                 1210.0  ...                   6541   \n",
       "...                                ...  ...                    ...   \n",
       "ENSG00000288721                  422.0  ...                     24   \n",
       "ENSG00000288722                  121.0  ...                      1   \n",
       "ENSG00000288723                    1.0  ...                      1   \n",
       "ENSG00000288724                    1.0  ...                     33   \n",
       "ENSG00000288725                    1.0  ...                    200   \n",
       "\n",
       "                 SRR12362564_Disease.2  SRR23920508_1_Disease  \\\n",
       "ENSG_ID                                                         \n",
       "ENSG00000000003                   2032                    0.0   \n",
       "ENSG00000000005                   2032                    0.0   \n",
       "ENSG00000000419                      2                    0.0   \n",
       "ENSG00000000457                      2                    0.0   \n",
       "ENSG00000000460                   6542                   10.0   \n",
       "...                                ...                    ...   \n",
       "ENSG00000288721                     25                    0.0   \n",
       "ENSG00000288722                      2                    0.0   \n",
       "ENSG00000288723                      2                    0.0   \n",
       "ENSG00000288724                     34                    0.0   \n",
       "ENSG00000288725                    201                    0.0   \n",
       "\n",
       "                 SRR23920508_1_Disease.1  SRR23920508_1_Disease.2  \\\n",
       "ENSG_ID                                                             \n",
       "ENSG00000000003                      1.0                      2.0   \n",
       "ENSG00000000005                      1.0                      2.0   \n",
       "ENSG00000000419                      1.0                      2.0   \n",
       "ENSG00000000457                      1.0                      2.0   \n",
       "ENSG00000000460                     11.0                     12.0   \n",
       "...                                  ...                      ...   \n",
       "ENSG00000288721                      1.0                      2.0   \n",
       "ENSG00000288722                      1.0                      2.0   \n",
       "ENSG00000288723                      1.0                      2.0   \n",
       "ENSG00000288724                      1.0                      2.0   \n",
       "ENSG00000288725                      1.0                      2.0   \n",
       "\n",
       "                 SRR23920508_1_Disease.3  SRR23920508_1_Disease.4  \\\n",
       "ENSG_ID                                                             \n",
       "ENSG00000000003                      3.0                      4.0   \n",
       "ENSG00000000005                      3.0                      4.0   \n",
       "ENSG00000000419                      3.0                      4.0   \n",
       "ENSG00000000457                      3.0                      4.0   \n",
       "ENSG00000000460                     13.0                     14.0   \n",
       "...                                  ...                      ...   \n",
       "ENSG00000288721                      3.0                      4.0   \n",
       "ENSG00000288722                      3.0                      4.0   \n",
       "ENSG00000288723                      3.0                      4.0   \n",
       "ENSG00000288724                      3.0                      4.0   \n",
       "ENSG00000288725                      3.0                      4.0   \n",
       "\n",
       "                 SRR23920508_1_Disease.5  SRR23920508_1_Disease.6  \\\n",
       "ENSG_ID                                                             \n",
       "ENSG00000000003                      5.0                      6.0   \n",
       "ENSG00000000005                      5.0                      6.0   \n",
       "ENSG00000000419                      5.0                      6.0   \n",
       "ENSG00000000457                      5.0                      6.0   \n",
       "ENSG00000000460                     15.0                     16.0   \n",
       "...                                  ...                      ...   \n",
       "ENSG00000288721                      5.0                      6.0   \n",
       "ENSG00000288722                      5.0                      6.0   \n",
       "ENSG00000288723                      5.0                      6.0   \n",
       "ENSG00000288724                      5.0                      6.0   \n",
       "ENSG00000288725                      5.0                      6.0   \n",
       "\n",
       "                 SRR23920508_1_Disease.7  \n",
       "ENSG_ID                                   \n",
       "ENSG00000000003                      7.0  \n",
       "ENSG00000000005                      7.0  \n",
       "ENSG00000000419                      7.0  \n",
       "ENSG00000000457                      7.0  \n",
       "ENSG00000000460                     17.0  \n",
       "...                                  ...  \n",
       "ENSG00000288721                      7.0  \n",
       "ENSG00000288722                      7.0  \n",
       "ENSG00000288723                      7.0  \n",
       "ENSG00000288724                      7.0  \n",
       "ENSG00000288725                      7.0  \n",
       "\n",
       "[60664 rows x 61 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts = counts.set_index('ENSG_ID')\n",
    "counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "80102b7e-ff6e-4435-997b-0edd97e5cc85",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts = counts.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a7fc75dc-96eb-4bcd-a1e7-2bf85a224251",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Group</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Samples</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR11235153_Normal</th>\n",
       "      <td>Normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR11235154_Normal</th>\n",
       "      <td>Normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR11235155_Disease</th>\n",
       "      <td>Diseased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR11235156_Disease</th>\n",
       "      <td>Diseased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR11235157_Disease</th>\n",
       "      <td>Diseased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23920508_1_Disease.3</th>\n",
       "      <td>Diseased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23920508_1_Disease.4</th>\n",
       "      <td>Diseased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23920508_1_Disease.5</th>\n",
       "      <td>Diseased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23920508_1_Disease.6</th>\n",
       "      <td>Diseased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR23920508_1_Disease.7</th>\n",
       "      <td>Diseased</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            Group\n",
       "Samples                          \n",
       "SRR11235153_Normal         Normal\n",
       "SRR11235154_Normal         Normal\n",
       "SRR11235155_Disease      Diseased\n",
       "SRR11235156_Disease      Diseased\n",
       "SRR11235157_Disease      Diseased\n",
       "...                           ...\n",
       "SRR23920508_1_Disease.3  Diseased\n",
       "SRR23920508_1_Disease.4  Diseased\n",
       "SRR23920508_1_Disease.5  Diseased\n",
       "SRR23920508_1_Disease.6  Diseased\n",
       "SRR23920508_1_Disease.7  Diseased\n",
       "\n",
       "[61 rows x 1 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata = pd.read_csv(\"annotation.csv\")\n",
    "metadata = metadata.set_index(\"Samples\")\n",
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1f4638b4-5047-4dc8-86c2-0a39601d7a9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "dds = DeseqDataSet(counts=counts,\n",
    "            metadata=metadata,\n",
    "            design_factors=\"Group\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ae430a2b-4d2b-48a5-8861-3f1f3ddd53d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.22 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 40.45 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 2.85 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 38.48 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 48.17 seconds.\n",
      "\n",
      "Replacing 3565 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 2.42 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 2.49 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 2.77 seconds.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "dds.deseq2()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ac2bea1f-0fa9-4afb-adbc-375485a0ea1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "stat_res = DeseqStats(dds,contrast = ('Group','Normal','Diseased'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d08fb826-58ed-438b-9d85-dcb67c862e5c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Running Wald tests...\n",
      "... done in 7.41 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Group Normal vs Diseased\n",
      "                    baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "ENSG_ID                                                                      \n",
      "ENSG00000000003  1194.077676        0.557220  0.792371  0.703231  0.481912   \n",
      "ENSG00000000005   287.281831       -0.571163  1.190087 -0.479934  0.631274   \n",
      "ENSG00000000419  1557.627891        0.105697  1.070422  0.098743  0.921342   \n",
      "ENSG00000000457   443.825278       -0.937375  0.953292 -0.983303  0.325458   \n",
      "ENSG00000000460   498.291326       -2.272232  1.044923 -2.174546  0.029664   \n",
      "...                      ...             ...       ...       ...       ...   \n",
      "ENSG00000288721    72.367904        0.303627  0.972559  0.312194  0.754893   \n",
      "ENSG00000288722   283.171737        0.059031  1.101170  0.053608  0.957248   \n",
      "ENSG00000288723    25.215058       -2.750390  2.037867 -1.349642  0.177131   \n",
      "ENSG00000288724    25.914691       -5.884235  2.950028 -1.994637  0.046082   \n",
      "ENSG00000288725    30.867923       -6.758629  3.066598 -2.203950  0.027528   \n",
      "\n",
      "                     padj  \n",
      "ENSG_ID                    \n",
      "ENSG00000000003  0.672828  \n",
      "ENSG00000000005  0.786183  \n",
      "ENSG00000000419  0.961973  \n",
      "ENSG00000000457  0.530048  \n",
      "ENSG00000000460  0.100378  \n",
      "...                   ...  \n",
      "ENSG00000288721  0.868583  \n",
      "ENSG00000288722  0.978366  \n",
      "ENSG00000288723       NaN  \n",
      "ENSG00000288724       NaN  \n",
      "ENSG00000288725       NaN  \n",
      "\n",
      "[60664 rows x 6 columns]\n"
     ]
    }
   ],
   "source": [
    "stat_res.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "de96a98c-687e-4511-9ea1-4b5fdf054bde",
   "metadata": {},
   "outputs": [],
   "source": [
    "res = stat_res.results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "6d274a23-62e6-4f77-b610-856e8c54b34a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>1194.077676</td>\n",
       "      <td>0.557220</td>\n",
       "      <td>0.792371</td>\n",
       "      <td>0.703231</td>\n",
       "      <td>0.481912</td>\n",
       "      <td>0.672828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005</th>\n",
       "      <td>287.281831</td>\n",
       "      <td>-0.571163</td>\n",
       "      <td>1.190087</td>\n",
       "      <td>-0.479934</td>\n",
       "      <td>0.631274</td>\n",
       "      <td>0.786183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>1557.627891</td>\n",
       "      <td>0.105697</td>\n",
       "      <td>1.070422</td>\n",
       "      <td>0.098743</td>\n",
       "      <td>0.921342</td>\n",
       "      <td>0.961973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <td>443.825278</td>\n",
       "      <td>-0.937375</td>\n",
       "      <td>0.953292</td>\n",
       "      <td>-0.983303</td>\n",
       "      <td>0.325458</td>\n",
       "      <td>0.530048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>498.291326</td>\n",
       "      <td>-2.272232</td>\n",
       "      <td>1.044923</td>\n",
       "      <td>-2.174546</td>\n",
       "      <td>0.029664</td>\n",
       "      <td>0.100378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288721</th>\n",
       "      <td>72.367904</td>\n",
       "      <td>0.303627</td>\n",
       "      <td>0.972559</td>\n",
       "      <td>0.312194</td>\n",
       "      <td>0.754893</td>\n",
       "      <td>0.868583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288722</th>\n",
       "      <td>283.171737</td>\n",
       "      <td>0.059031</td>\n",
       "      <td>1.101170</td>\n",
       "      <td>0.053608</td>\n",
       "      <td>0.957248</td>\n",
       "      <td>0.978366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288723</th>\n",
       "      <td>25.215058</td>\n",
       "      <td>-2.750390</td>\n",
       "      <td>2.037867</td>\n",
       "      <td>-1.349642</td>\n",
       "      <td>0.177131</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288724</th>\n",
       "      <td>25.914691</td>\n",
       "      <td>-5.884235</td>\n",
       "      <td>2.950028</td>\n",
       "      <td>-1.994637</td>\n",
       "      <td>0.046082</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288725</th>\n",
       "      <td>30.867923</td>\n",
       "      <td>-6.758629</td>\n",
       "      <td>3.066598</td>\n",
       "      <td>-2.203950</td>\n",
       "      <td>0.027528</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60664 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
       "ENSG_ID                                                                      \n",
       "ENSG00000000003  1194.077676        0.557220  0.792371  0.703231  0.481912   \n",
       "ENSG00000000005   287.281831       -0.571163  1.190087 -0.479934  0.631274   \n",
       "ENSG00000000419  1557.627891        0.105697  1.070422  0.098743  0.921342   \n",
       "ENSG00000000457   443.825278       -0.937375  0.953292 -0.983303  0.325458   \n",
       "ENSG00000000460   498.291326       -2.272232  1.044923 -2.174546  0.029664   \n",
       "...                      ...             ...       ...       ...       ...   \n",
       "ENSG00000288721    72.367904        0.303627  0.972559  0.312194  0.754893   \n",
       "ENSG00000288722   283.171737        0.059031  1.101170  0.053608  0.957248   \n",
       "ENSG00000288723    25.215058       -2.750390  2.037867 -1.349642  0.177131   \n",
       "ENSG00000288724    25.914691       -5.884235  2.950028 -1.994637  0.046082   \n",
       "ENSG00000288725    30.867923       -6.758629  3.066598 -2.203950  0.027528   \n",
       "\n",
       "                     padj  \n",
       "ENSG_ID                    \n",
       "ENSG00000000003  0.672828  \n",
       "ENSG00000000005  0.786183  \n",
       "ENSG00000000419  0.961973  \n",
       "ENSG00000000457  0.530048  \n",
       "ENSG00000000460  0.100378  \n",
       "...                   ...  \n",
       "ENSG00000288721  0.868583  \n",
       "ENSG00000288722  0.978366  \n",
       "ENSG00000288723       NaN  \n",
       "ENSG00000288724       NaN  \n",
       "ENSG00000288725       NaN  \n",
       "\n",
       "[60664 rows x 6 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "8341bdc2-3415-44bc-b1de-e43b5fb2192c",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapper = id_map(species = 'human')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "92e5cbc3-91c8-4681-9f61-721039a4f5fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "res['Symbol'] = res.index.map(mapper.mapper) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "4038ac5f-97a7-48f3-bda0-e5df34c4b8b9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>Symbol</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>1194.077676</td>\n",
       "      <td>0.557220</td>\n",
       "      <td>0.792371</td>\n",
       "      <td>0.703231</td>\n",
       "      <td>0.481912</td>\n",
       "      <td>0.672828</td>\n",
       "      <td>TSPAN6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005</th>\n",
       "      <td>287.281831</td>\n",
       "      <td>-0.571163</td>\n",
       "      <td>1.190087</td>\n",
       "      <td>-0.479934</td>\n",
       "      <td>0.631274</td>\n",
       "      <td>0.786183</td>\n",
       "      <td>TNMD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>1557.627891</td>\n",
       "      <td>0.105697</td>\n",
       "      <td>1.070422</td>\n",
       "      <td>0.098743</td>\n",
       "      <td>0.921342</td>\n",
       "      <td>0.961973</td>\n",
       "      <td>DPM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <td>443.825278</td>\n",
       "      <td>-0.937375</td>\n",
       "      <td>0.953292</td>\n",
       "      <td>-0.983303</td>\n",
       "      <td>0.325458</td>\n",
       "      <td>0.530048</td>\n",
       "      <td>SCYL3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>498.291326</td>\n",
       "      <td>-2.272232</td>\n",
       "      <td>1.044923</td>\n",
       "      <td>-2.174546</td>\n",
       "      <td>0.029664</td>\n",
       "      <td>0.100378</td>\n",
       "      <td>C1orf112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288721</th>\n",
       "      <td>72.367904</td>\n",
       "      <td>0.303627</td>\n",
       "      <td>0.972559</td>\n",
       "      <td>0.312194</td>\n",
       "      <td>0.754893</td>\n",
       "      <td>0.868583</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288722</th>\n",
       "      <td>283.171737</td>\n",
       "      <td>0.059031</td>\n",
       "      <td>1.101170</td>\n",
       "      <td>0.053608</td>\n",
       "      <td>0.957248</td>\n",
       "      <td>0.978366</td>\n",
       "      <td>F8A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288723</th>\n",
       "      <td>25.215058</td>\n",
       "      <td>-2.750390</td>\n",
       "      <td>2.037867</td>\n",
       "      <td>-1.349642</td>\n",
       "      <td>0.177131</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288724</th>\n",
       "      <td>25.914691</td>\n",
       "      <td>-5.884235</td>\n",
       "      <td>2.950028</td>\n",
       "      <td>-1.994637</td>\n",
       "      <td>0.046082</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288725</th>\n",
       "      <td>30.867923</td>\n",
       "      <td>-6.758629</td>\n",
       "      <td>3.066598</td>\n",
       "      <td>-2.203950</td>\n",
       "      <td>0.027528</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60664 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
       "ENSG_ID                                                                      \n",
       "ENSG00000000003  1194.077676        0.557220  0.792371  0.703231  0.481912   \n",
       "ENSG00000000005   287.281831       -0.571163  1.190087 -0.479934  0.631274   \n",
       "ENSG00000000419  1557.627891        0.105697  1.070422  0.098743  0.921342   \n",
       "ENSG00000000457   443.825278       -0.937375  0.953292 -0.983303  0.325458   \n",
       "ENSG00000000460   498.291326       -2.272232  1.044923 -2.174546  0.029664   \n",
       "...                      ...             ...       ...       ...       ...   \n",
       "ENSG00000288721    72.367904        0.303627  0.972559  0.312194  0.754893   \n",
       "ENSG00000288722   283.171737        0.059031  1.101170  0.053608  0.957248   \n",
       "ENSG00000288723    25.215058       -2.750390  2.037867 -1.349642  0.177131   \n",
       "ENSG00000288724    25.914691       -5.884235  2.950028 -1.994637  0.046082   \n",
       "ENSG00000288725    30.867923       -6.758629  3.066598 -2.203950  0.027528   \n",
       "\n",
       "                     padj    Symbol  \n",
       "ENSG_ID                              \n",
       "ENSG00000000003  0.672828    TSPAN6  \n",
       "ENSG00000000005  0.786183      TNMD  \n",
       "ENSG00000000419  0.961973      DPM1  \n",
       "ENSG00000000457  0.530048     SCYL3  \n",
       "ENSG00000000460  0.100378  C1orf112  \n",
       "...                   ...       ...  \n",
       "ENSG00000288721  0.868583       NaN  \n",
       "ENSG00000288722  0.978366      F8A1  \n",
       "ENSG00000288723       NaN       NaN  \n",
       "ENSG00000288724       NaN       NaN  \n",
       "ENSG00000288725       NaN       NaN  \n",
       "\n",
       "[60664 rows x 7 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "1ac70399-a514-4866-a293-116607890570",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "879dbd4d-4775-4a65-bd6a-916a46562008",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.rename(columns = {'Symbol' : 'Gene'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "dd6c87f9-7971-46a8-9a93-756b8ee95021",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>Gene</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>1194.077676</td>\n",
       "      <td>0.557220</td>\n",
       "      <td>0.792371</td>\n",
       "      <td>0.703231</td>\n",
       "      <td>0.481912</td>\n",
       "      <td>0.672828</td>\n",
       "      <td>TSPAN6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005</th>\n",
       "      <td>287.281831</td>\n",
       "      <td>-0.571163</td>\n",
       "      <td>1.190087</td>\n",
       "      <td>-0.479934</td>\n",
       "      <td>0.631274</td>\n",
       "      <td>0.786183</td>\n",
       "      <td>TNMD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>1557.627891</td>\n",
       "      <td>0.105697</td>\n",
       "      <td>1.070422</td>\n",
       "      <td>0.098743</td>\n",
       "      <td>0.921342</td>\n",
       "      <td>0.961973</td>\n",
       "      <td>DPM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <td>443.825278</td>\n",
       "      <td>-0.937375</td>\n",
       "      <td>0.953292</td>\n",
       "      <td>-0.983303</td>\n",
       "      <td>0.325458</td>\n",
       "      <td>0.530048</td>\n",
       "      <td>SCYL3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>498.291326</td>\n",
       "      <td>-2.272232</td>\n",
       "      <td>1.044923</td>\n",
       "      <td>-2.174546</td>\n",
       "      <td>0.029664</td>\n",
       "      <td>0.100378</td>\n",
       "      <td>C1orf112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288721</th>\n",
       "      <td>72.367904</td>\n",
       "      <td>0.303627</td>\n",
       "      <td>0.972559</td>\n",
       "      <td>0.312194</td>\n",
       "      <td>0.754893</td>\n",
       "      <td>0.868583</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288722</th>\n",
       "      <td>283.171737</td>\n",
       "      <td>0.059031</td>\n",
       "      <td>1.101170</td>\n",
       "      <td>0.053608</td>\n",
       "      <td>0.957248</td>\n",
       "      <td>0.978366</td>\n",
       "      <td>F8A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288723</th>\n",
       "      <td>25.215058</td>\n",
       "      <td>-2.750390</td>\n",
       "      <td>2.037867</td>\n",
       "      <td>-1.349642</td>\n",
       "      <td>0.177131</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288724</th>\n",
       "      <td>25.914691</td>\n",
       "      <td>-5.884235</td>\n",
       "      <td>2.950028</td>\n",
       "      <td>-1.994637</td>\n",
       "      <td>0.046082</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288725</th>\n",
       "      <td>30.867923</td>\n",
       "      <td>-6.758629</td>\n",
       "      <td>3.066598</td>\n",
       "      <td>-2.203950</td>\n",
       "      <td>0.027528</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60664 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
       "ENSG_ID                                                                      \n",
       "ENSG00000000003  1194.077676        0.557220  0.792371  0.703231  0.481912   \n",
       "ENSG00000000005   287.281831       -0.571163  1.190087 -0.479934  0.631274   \n",
       "ENSG00000000419  1557.627891        0.105697  1.070422  0.098743  0.921342   \n",
       "ENSG00000000457   443.825278       -0.937375  0.953292 -0.983303  0.325458   \n",
       "ENSG00000000460   498.291326       -2.272232  1.044923 -2.174546  0.029664   \n",
       "...                      ...             ...       ...       ...       ...   \n",
       "ENSG00000288721    72.367904        0.303627  0.972559  0.312194  0.754893   \n",
       "ENSG00000288722   283.171737        0.059031  1.101170  0.053608  0.957248   \n",
       "ENSG00000288723    25.215058       -2.750390  2.037867 -1.349642  0.177131   \n",
       "ENSG00000288724    25.914691       -5.884235  2.950028 -1.994637  0.046082   \n",
       "ENSG00000288725    30.867923       -6.758629  3.066598 -2.203950  0.027528   \n",
       "\n",
       "                     padj      Gene  \n",
       "ENSG_ID                              \n",
       "ENSG00000000003  0.672828    TSPAN6  \n",
       "ENSG00000000005  0.786183      TNMD  \n",
       "ENSG00000000419  0.961973      DPM1  \n",
       "ENSG00000000457  0.530048     SCYL3  \n",
       "ENSG00000000460  0.100378  C1orf112  \n",
       "...                   ...       ...  \n",
       "ENSG00000288721  0.868583       NaN  \n",
       "ENSG00000288722  0.978366      F8A1  \n",
       "ENSG00000288723       NaN       NaN  \n",
       "ENSG00000288724       NaN       NaN  \n",
       "ENSG00000288725       NaN       NaN  \n",
       "\n",
       "[60664 rows x 7 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "d69798c3-9ded-458f-9096-87f74c630c6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "8bcba6c8-d3a5-435e-adaf-a6880164e730",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['Rank'] = -np.log10(df.padj)*df.log2FoldChange"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "089039bf-f6cb-423f-96a3-1e962229d733",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.sort_values('Rank', ascending = False).reset_index(drop = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "6c61ba97-ebdf-45fd-9d9e-92111d2d2532",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11455.955249</td>\n",
       "      <td>11.619298</td>\n",
       "      <td>1.081715</td>\n",
       "      <td>10.741552</td>\n",
       "      <td>6.494343e-27</td>\n",
       "      <td>1.877385e-22</td>\n",
       "      <td>MYL4</td>\n",
       "      <td>252.446069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3857.588456</td>\n",
       "      <td>10.124764</td>\n",
       "      <td>1.372634</td>\n",
       "      <td>7.376157</td>\n",
       "      <td>1.629241e-13</td>\n",
       "      <td>5.887263e-10</td>\n",
       "      <td>NPPA</td>\n",
       "      <td>93.452449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>273.403272</td>\n",
       "      <td>7.741716</td>\n",
       "      <td>0.965341</td>\n",
       "      <td>8.019673</td>\n",
       "      <td>1.060274e-15</td>\n",
       "      <td>7.662602e-12</td>\n",
       "      <td>AC104463.2</td>\n",
       "      <td>86.053998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>832.254601</td>\n",
       "      <td>6.354969</td>\n",
       "      <td>0.874496</td>\n",
       "      <td>7.267006</td>\n",
       "      <td>3.675418e-13</td>\n",
       "      <td>9.806771e-10</td>\n",
       "      <td>QPRT</td>\n",
       "      <td>57.248574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>438.782464</td>\n",
       "      <td>5.282494</td>\n",
       "      <td>0.689224</td>\n",
       "      <td>7.664403</td>\n",
       "      <td>1.796656e-14</td>\n",
       "      <td>1.038755e-10</td>\n",
       "      <td>BAMBI</td>\n",
       "      <td>52.737705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60659</th>\n",
       "      <td>26.021652</td>\n",
       "      <td>-6.015443</td>\n",
       "      <td>3.026929</td>\n",
       "      <td>-1.987309</td>\n",
       "      <td>4.688815e-02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60660</th>\n",
       "      <td>24.438538</td>\n",
       "      <td>-5.606788</td>\n",
       "      <td>3.074482</td>\n",
       "      <td>-1.823653</td>\n",
       "      <td>6.820460e-02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60661</th>\n",
       "      <td>25.215058</td>\n",
       "      <td>-2.750390</td>\n",
       "      <td>2.037867</td>\n",
       "      <td>-1.349642</td>\n",
       "      <td>1.771309e-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60662</th>\n",
       "      <td>25.914691</td>\n",
       "      <td>-5.884235</td>\n",
       "      <td>2.950028</td>\n",
       "      <td>-1.994637</td>\n",
       "      <td>4.608244e-02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60663</th>\n",
       "      <td>30.867923</td>\n",
       "      <td>-6.758629</td>\n",
       "      <td>3.066598</td>\n",
       "      <td>-2.203950</td>\n",
       "      <td>2.752785e-02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60664 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           baseMean  log2FoldChange     lfcSE       stat        pvalue  \\\n",
       "0      11455.955249       11.619298  1.081715  10.741552  6.494343e-27   \n",
       "1       3857.588456       10.124764  1.372634   7.376157  1.629241e-13   \n",
       "2        273.403272        7.741716  0.965341   8.019673  1.060274e-15   \n",
       "3        832.254601        6.354969  0.874496   7.267006  3.675418e-13   \n",
       "4        438.782464        5.282494  0.689224   7.664403  1.796656e-14   \n",
       "...             ...             ...       ...        ...           ...   \n",
       "60659     26.021652       -6.015443  3.026929  -1.987309  4.688815e-02   \n",
       "60660     24.438538       -5.606788  3.074482  -1.823653  6.820460e-02   \n",
       "60661     25.215058       -2.750390  2.037867  -1.349642  1.771309e-01   \n",
       "60662     25.914691       -5.884235  2.950028  -1.994637  4.608244e-02   \n",
       "60663     30.867923       -6.758629  3.066598  -2.203950  2.752785e-02   \n",
       "\n",
       "               padj        Gene        Rank  \n",
       "0      1.877385e-22        MYL4  252.446069  \n",
       "1      5.887263e-10        NPPA   93.452449  \n",
       "2      7.662602e-12  AC104463.2   86.053998  \n",
       "3      9.806771e-10        QPRT   57.248574  \n",
       "4      1.038755e-10       BAMBI   52.737705  \n",
       "...             ...         ...         ...  \n",
       "60659           NaN         NaN         NaN  \n",
       "60660           NaN         NaN         NaN  \n",
       "60661           NaN         NaN         NaN  \n",
       "60662           NaN         NaN         NaN  \n",
       "60663           NaN         NaN         NaN  \n",
       "\n",
       "[60664 rows x 8 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "fdce79bc-b975-4704-8531-fc449317b5a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "ranking = df[['Gene', 'Rank']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "52ccb378-f2e7-4c13-ae29-da71caa3cdb5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MYL4</td>\n",
       "      <td>252.446069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NPPA</td>\n",
       "      <td>93.452449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AC104463.2</td>\n",
       "      <td>86.053998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>QPRT</td>\n",
       "      <td>57.248574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BAMBI</td>\n",
       "      <td>52.737705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60659</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60660</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60661</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60662</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60663</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60664 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Gene        Rank\n",
       "0            MYL4  252.446069\n",
       "1            NPPA   93.452449\n",
       "2      AC104463.2   86.053998\n",
       "3            QPRT   57.248574\n",
       "4           BAMBI   52.737705\n",
       "...           ...         ...\n",
       "60659         NaN         NaN\n",
       "60660         NaN         NaN\n",
       "60661         NaN         NaN\n",
       "60662         NaN         NaN\n",
       "60663         NaN         NaN\n",
       "\n",
       "[60664 rows x 2 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ranking"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "f36ef4be-f9fa-484d-a1ab-62b5fd571dda",
   "metadata": {},
   "outputs": [],
   "source": [
    "ranking = ranking.dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "9b171f28-313f-45af-9434-049be52337ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MYL4</td>\n",
       "      <td>252.446069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NPPA</td>\n",
       "      <td>93.452449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AC104463.2</td>\n",
       "      <td>86.053998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>QPRT</td>\n",
       "      <td>57.248574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BAMBI</td>\n",
       "      <td>52.737705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28903</th>\n",
       "      <td>EMP1</td>\n",
       "      <td>-106.766130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28904</th>\n",
       "      <td>EFEMP1</td>\n",
       "      <td>-108.354066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28905</th>\n",
       "      <td>SRGN</td>\n",
       "      <td>-108.466589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28906</th>\n",
       "      <td>PNPO</td>\n",
       "      <td>-114.940424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28907</th>\n",
       "      <td>VWF</td>\n",
       "      <td>-128.040136</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>26882 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Gene        Rank\n",
       "0            MYL4  252.446069\n",
       "1            NPPA   93.452449\n",
       "2      AC104463.2   86.053998\n",
       "3            QPRT   57.248574\n",
       "4           BAMBI   52.737705\n",
       "...           ...         ...\n",
       "28903        EMP1 -106.766130\n",
       "28904      EFEMP1 -108.354066\n",
       "28905        SRGN -108.466589\n",
       "28906        PNPO -114.940424\n",
       "28907         VWF -128.040136\n",
       "\n",
       "[26882 rows x 2 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ranking"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "b98d75cb-5a4e-406b-8535-0b423ce6e894",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ARCHS4_Cell-lines',\n",
       " 'ARCHS4_IDG_Coexp',\n",
       " 'ARCHS4_Kinases_Coexp',\n",
       " 'ARCHS4_TFs_Coexp',\n",
       " 'ARCHS4_Tissues',\n",
       " 'Achilles_fitness_decrease',\n",
       " 'Achilles_fitness_increase',\n",
       " 'Aging_Perturbations_from_GEO_down',\n",
       " 'Aging_Perturbations_from_GEO_up',\n",
       " 'Allen_Brain_Atlas_10x_scRNA_2021',\n",
       " 'Allen_Brain_Atlas_down',\n",
       " 'Allen_Brain_Atlas_up',\n",
       " 'Azimuth_2023',\n",
       " 'Azimuth_Cell_Types_2021',\n",
       " 'BioCarta_2013',\n",
       " 'BioCarta_2015',\n",
       " 'BioCarta_2016',\n",
       " 'BioPlanet_2019',\n",
       " 'BioPlex_2017',\n",
       " 'CCLE_Proteomics_2020',\n",
       " 'CORUM',\n",
       " 'COVID-19_Related_Gene_Sets',\n",
       " 'COVID-19_Related_Gene_Sets_2021',\n",
       " 'Cancer_Cell_Line_Encyclopedia',\n",
       " 'CellMarker_2024',\n",
       " 'CellMarker_Augmented_2021',\n",
       " 'ChEA_2013',\n",
       " 'ChEA_2015',\n",
       " 'ChEA_2016',\n",
       " 'ChEA_2022',\n",
       " 'Chromosome_Location',\n",
       " 'Chromosome_Location_hg19',\n",
       " 'ClinVar_2019',\n",
       " 'DSigDB',\n",
       " 'Data_Acquisition_Method_Most_Popular_Genes',\n",
       " 'DepMap_WG_CRISPR_Screens_Broad_CellLines_2019',\n",
       " 'DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019',\n",
       " 'Descartes_Cell_Types_and_Tissue_2021',\n",
       " 'Diabetes_Perturbations_GEO_2022',\n",
       " 'DisGeNET',\n",
       " 'Disease_Perturbations_from_GEO_down',\n",
       " 'Disease_Perturbations_from_GEO_up',\n",
       " 'Disease_Signatures_from_GEO_down_2014',\n",
       " 'Disease_Signatures_from_GEO_up_2014',\n",
       " 'DrugMatrix',\n",
       " 'Drug_Perturbations_from_GEO_2014',\n",
       " 'Drug_Perturbations_from_GEO_down',\n",
       " 'Drug_Perturbations_from_GEO_up',\n",
       " 'ENCODE_Histone_Modifications_2013',\n",
       " 'ENCODE_Histone_Modifications_2015',\n",
       " 'ENCODE_TF_ChIP-seq_2014',\n",
       " 'ENCODE_TF_ChIP-seq_2015',\n",
       " 'ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X',\n",
       " 'ESCAPE',\n",
       " 'Elsevier_Pathway_Collection',\n",
       " 'Enrichr_Libraries_Most_Popular_Genes',\n",
       " 'Enrichr_Submissions_TF-Gene_Coocurrence',\n",
       " 'Enrichr_Users_Contributed_Lists_2020',\n",
       " 'Epigenomics_Roadmap_HM_ChIP-seq',\n",
       " 'FANTOM6_lncRNA_KD_DEGs',\n",
       " 'GO_Biological_Process_2013',\n",
       " 'GO_Biological_Process_2015',\n",
       " 'GO_Biological_Process_2017',\n",
       " 'GO_Biological_Process_2017b',\n",
       " 'GO_Biological_Process_2018',\n",
       " 'GO_Biological_Process_2021',\n",
       " 'GO_Biological_Process_2023',\n",
       " 'GO_Cellular_Component_2013',\n",
       " 'GO_Cellular_Component_2015',\n",
       " 'GO_Cellular_Component_2017',\n",
       " 'GO_Cellular_Component_2017b',\n",
       " 'GO_Cellular_Component_2018',\n",
       " 'GO_Cellular_Component_2021',\n",
       " 'GO_Cellular_Component_2023',\n",
       " 'GO_Molecular_Function_2013',\n",
       " 'GO_Molecular_Function_2015',\n",
       " 'GO_Molecular_Function_2017',\n",
       " 'GO_Molecular_Function_2017b',\n",
       " 'GO_Molecular_Function_2018',\n",
       " 'GO_Molecular_Function_2021',\n",
       " 'GO_Molecular_Function_2023',\n",
       " 'GTEx_Aging_Signatures_2021',\n",
       " 'GTEx_Tissue_Expression_Down',\n",
       " 'GTEx_Tissue_Expression_Up',\n",
       " 'GTEx_Tissues_V8_2023',\n",
       " 'GWAS_Catalog_2019',\n",
       " 'GWAS_Catalog_2023',\n",
       " 'GeDiPNet_2023',\n",
       " 'GeneSigDB',\n",
       " 'Gene_Perturbations_from_GEO_down',\n",
       " 'Gene_Perturbations_from_GEO_up',\n",
       " 'Genes_Associated_with_NIH_Grants',\n",
       " 'Genome_Browser_PWMs',\n",
       " 'GlyGen_Glycosylated_Proteins_2022',\n",
       " 'HDSigDB_Human_2021',\n",
       " 'HDSigDB_Mouse_2021',\n",
       " 'HMDB_Metabolites',\n",
       " 'HMS_LINCS_KinomeScan',\n",
       " 'HomoloGene',\n",
       " 'HuBMAP_ASCT_plus_B_augmented_w_RNAseq_Coexpression',\n",
       " 'HuBMAP_ASCTplusB_augmented_2022',\n",
       " 'HumanCyc_2015',\n",
       " 'HumanCyc_2016',\n",
       " 'Human_Gene_Atlas',\n",
       " 'Human_Phenotype_Ontology',\n",
       " 'IDG_Drug_Targets_2022',\n",
       " 'InterPro_Domains_2019',\n",
       " 'Jensen_COMPARTMENTS',\n",
       " 'Jensen_DISEASES',\n",
       " 'Jensen_TISSUES',\n",
       " 'KEA_2013',\n",
       " 'KEA_2015',\n",
       " 'KEGG_2013',\n",
       " 'KEGG_2015',\n",
       " 'KEGG_2016',\n",
       " 'KEGG_2019_Human',\n",
       " 'KEGG_2019_Mouse',\n",
       " 'KEGG_2021_Human',\n",
       " 'KOMP2_Mouse_Phenotypes_2022',\n",
       " 'Kinase_Perturbations_from_GEO_down',\n",
       " 'Kinase_Perturbations_from_GEO_up',\n",
       " 'L1000_Kinase_and_GPCR_Perturbations_down',\n",
       " 'L1000_Kinase_and_GPCR_Perturbations_up',\n",
       " 'LINCS_L1000_CRISPR_KO_Consensus_Sigs',\n",
       " 'LINCS_L1000_Chem_Pert_Consensus_Sigs',\n",
       " 'LINCS_L1000_Chem_Pert_down',\n",
       " 'LINCS_L1000_Chem_Pert_up',\n",
       " 'LINCS_L1000_Ligand_Perturbations_down',\n",
       " 'LINCS_L1000_Ligand_Perturbations_up',\n",
       " 'Ligand_Perturbations_from_GEO_down',\n",
       " 'Ligand_Perturbations_from_GEO_up',\n",
       " 'MAGMA_Drugs_and_Diseases',\n",
       " 'MAGNET_2023',\n",
       " 'MCF7_Perturbations_from_GEO_down',\n",
       " 'MCF7_Perturbations_from_GEO_up',\n",
       " 'MGI_Mammalian_Phenotype_2013',\n",
       " 'MGI_Mammalian_Phenotype_2017',\n",
       " 'MGI_Mammalian_Phenotype_Level_3',\n",
       " 'MGI_Mammalian_Phenotype_Level_4',\n",
       " 'MGI_Mammalian_Phenotype_Level_4_2019',\n",
       " 'MGI_Mammalian_Phenotype_Level_4_2021',\n",
       " 'MSigDB_Computational',\n",
       " 'MSigDB_Hallmark_2020',\n",
       " 'MSigDB_Oncogenic_Signatures',\n",
       " 'Metabolomics_Workbench_Metabolites_2022',\n",
       " 'Microbe_Perturbations_from_GEO_down',\n",
       " 'Microbe_Perturbations_from_GEO_up',\n",
       " 'MoTrPAC_2023',\n",
       " 'Mouse_Gene_Atlas',\n",
       " 'NCI-60_Cancer_Cell_Lines',\n",
       " 'NCI-Nature_2015',\n",
       " 'NCI-Nature_2016',\n",
       " 'NIH_Funded_PIs_2017_AutoRIF_ARCHS4_Predictions',\n",
       " 'NIH_Funded_PIs_2017_GeneRIF_ARCHS4_Predictions',\n",
       " 'NIH_Funded_PIs_2017_Human_AutoRIF',\n",
       " 'NIH_Funded_PIs_2017_Human_GeneRIF',\n",
       " 'NURSA_Human_Endogenous_Complexome',\n",
       " 'OMIM_Disease',\n",
       " 'OMIM_Expanded',\n",
       " 'Old_CMAP_down',\n",
       " 'Old_CMAP_up',\n",
       " 'Orphanet_Augmented_2021',\n",
       " 'PFOCR_Pathways',\n",
       " 'PFOCR_Pathways_2023',\n",
       " 'PPI_Hub_Proteins',\n",
       " 'PanglaoDB_Augmented_2021',\n",
       " 'Panther_2015',\n",
       " 'Panther_2016',\n",
       " 'Pfam_Domains_2019',\n",
       " 'Pfam_InterPro_Domains',\n",
       " 'PheWeb_2019',\n",
       " 'PhenGenI_Association_2021',\n",
       " 'Phosphatase_Substrates_from_DEPOD',\n",
       " 'ProteomicsDB_2020',\n",
       " 'Proteomics_Drug_Atlas_2023',\n",
       " 'RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO',\n",
       " 'RNAseq_Automatic_GEO_Signatures_Human_Down',\n",
       " 'RNAseq_Automatic_GEO_Signatures_Human_Up',\n",
       " 'RNAseq_Automatic_GEO_Signatures_Mouse_Down',\n",
       " 'RNAseq_Automatic_GEO_Signatures_Mouse_Up',\n",
       " 'Rare_Diseases_AutoRIF_ARCHS4_Predictions',\n",
       " 'Rare_Diseases_AutoRIF_Gene_Lists',\n",
       " 'Rare_Diseases_GeneRIF_ARCHS4_Predictions',\n",
       " 'Rare_Diseases_GeneRIF_Gene_Lists',\n",
       " 'Reactome_2013',\n",
       " 'Reactome_2015',\n",
       " 'Reactome_2016',\n",
       " 'Reactome_2022',\n",
       " 'Rummagene_kinases',\n",
       " 'Rummagene_signatures',\n",
       " 'Rummagene_transcription_factors',\n",
       " 'SILAC_Phosphoproteomics',\n",
       " 'SubCell_BarCode',\n",
       " 'SynGO_2022',\n",
       " 'SynGO_2024',\n",
       " 'SysMyo_Muscle_Gene_Sets',\n",
       " 'TF-LOF_Expression_from_GEO',\n",
       " 'TF_Perturbations_Followed_by_Expression',\n",
       " 'TG_GATES_2020',\n",
       " 'TRANSFAC_and_JASPAR_PWMs',\n",
       " 'TRRUST_Transcription_Factors_2019',\n",
       " 'Table_Mining_of_CRISPR_Studies',\n",
       " 'Tabula_Muris',\n",
       " 'Tabula_Sapiens',\n",
       " 'TargetScan_microRNA',\n",
       " 'TargetScan_microRNA_2017',\n",
       " 'The_Kinase_Library_2023',\n",
       " 'Tissue_Protein_Expression_from_Human_Proteome_Map',\n",
       " 'Tissue_Protein_Expression_from_ProteomicsDB',\n",
       " 'Transcription_Factor_PPIs',\n",
       " 'UK_Biobank_GWAS_v1',\n",
       " 'Virus-Host_PPI_P-HIPSTer_2020',\n",
       " 'VirusMINT',\n",
       " 'Virus_Perturbations_from_GEO_down',\n",
       " 'Virus_Perturbations_from_GEO_up',\n",
       " 'WikiPathway_2021_Human',\n",
       " 'WikiPathway_2023_Human',\n",
       " 'WikiPathways_2013',\n",
       " 'WikiPathways_2015',\n",
       " 'WikiPathways_2016',\n",
       " 'WikiPathways_2019_Human',\n",
       " 'WikiPathways_2019_Mouse',\n",
       " 'dbGaP',\n",
       " 'huMAP',\n",
       " 'lncHUB_lncRNA_Co-Expression',\n",
       " 'miRTarBase_2017']"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gp.get_library_name()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "5074b267-26fd-4998-81cc-7af55b10098e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-04-29 00:12:41,781 [WARNING] Duplicated values found in preranked stats: 0.50% of genes\n",
      "The order of those genes will be arbitrary, which may produce unexpected results.\n"
     ]
    }
   ],
   "source": [
    "pre_res = gp.prerank(rnk = ranking, gene_sets = 'GO_Biological_Process_2021', seed = 6, permutation_num = 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "fe58205b-5615-444b-91ad-0d290751b3a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Term</th>\n",
       "      <th>fdr</th>\n",
       "      <th>es</th>\n",
       "      <th>nes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>regulation of microtubule depolymerization (GO...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.739856</td>\n",
       "      <td>2.092405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>response to muscle stretch (GO:0035994)</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.843178</td>\n",
       "      <td>2.610499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cristae formation (GO:0042407)</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.695494</td>\n",
       "      <td>2.268565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>negative regulation of microtubule depolymeriz...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.816338</td>\n",
       "      <td>1.979629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mitochondrial ATP synthesis coupled proton tra...</td>\n",
       "      <td>0.013054</td>\n",
       "      <td>0.811394</td>\n",
       "      <td>1.943826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2803</th>\n",
       "      <td>RNA-dependent DNA biosynthetic process (GO:000...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.579026</td>\n",
       "      <td>-0.924007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2804</th>\n",
       "      <td>C21-steroid hormone metabolic process (GO:0008...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.595209</td>\n",
       "      <td>-0.927550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2805</th>\n",
       "      <td>regulation of microtubule-based process (GO:00...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.376291</td>\n",
       "      <td>-0.590667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2806</th>\n",
       "      <td>embryonic digestive tract development (GO:0048...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.613664</td>\n",
       "      <td>-0.935118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2807</th>\n",
       "      <td>regulation of transcription from RNA polymeras...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.504141</td>\n",
       "      <td>-0.809870</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2808 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   Term       fdr        es  \\\n",
       "0     regulation of microtubule depolymerization (GO...  0.000000  0.739856   \n",
       "1               response to muscle stretch (GO:0035994)  0.000000  0.843178   \n",
       "2                        cristae formation (GO:0042407)  0.000000  0.695494   \n",
       "3     negative regulation of microtubule depolymeriz...  0.000000  0.816338   \n",
       "4     mitochondrial ATP synthesis coupled proton tra...  0.013054  0.811394   \n",
       "...                                                 ...       ...       ...   \n",
       "2803  RNA-dependent DNA biosynthetic process (GO:000...  1.000000 -0.579026   \n",
       "2804  C21-steroid hormone metabolic process (GO:0008...  1.000000 -0.595209   \n",
       "2805  regulation of microtubule-based process (GO:00...  1.000000 -0.376291   \n",
       "2806  embryonic digestive tract development (GO:0048...  1.000000 -0.613664   \n",
       "2807  regulation of transcription from RNA polymeras...  1.000000 -0.504141   \n",
       "\n",
       "           nes  \n",
       "0     2.092405  \n",
       "1     2.610499  \n",
       "2     2.268565  \n",
       "3     1.979629  \n",
       "4     1.943826  \n",
       "...        ...  \n",
       "2803 -0.924007  \n",
       "2804 -0.927550  \n",
       "2805 -0.590667  \n",
       "2806 -0.935118  \n",
       "2807 -0.809870  \n",
       "\n",
       "[2808 rows x 4 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out = []\n",
    "\n",
    "for term in list(pre_res.results):\n",
    "    out.append([term,\n",
    "               pre_res.results[term]['fdr'],\n",
    "               pre_res.results[term]['es'],\n",
    "               pre_res.results[term]['nes']])\n",
    "\n",
    "out_df = pd.DataFrame(out, columns = ['Term','fdr', 'es', 'nes']).sort_values('fdr').reset_index(drop = True)\n",
    "out_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "34fd0c5c-473c-4419-82cd-961a0c144d7f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'regulation of microtubule depolymerization (GO:0031114)'"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_df.loc[0]['Term']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "fa38beb2-9b4a-42e9-8cb7-991ea34fd713",
   "metadata": {},
   "outputs": [
    {
     "ename": "LookupError",
     "evalue": "No supported database. Please input one of these:\n                            ('Human', 'Mouse', 'Yeast', 'Fly', 'Fish', 'Worm') ",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mLookupError\u001b[0m                               Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[40], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m gene_sets \u001b[38;5;241m=\u001b[39m \u001b[43mgp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_library_name\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mMsigDb\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[0;32m      3\u001b[0m gene_set \u001b[38;5;241m=\u001b[39m gene_sets[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mregulation of microtubule depolymerization (GO:0031114)\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[0;32m      4\u001b[0m genes \u001b[38;5;241m=\u001b[39m gene_set\u001b[38;5;241m.\u001b[39mgenes\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python312\\site-packages\\gseapy\\parser.py:216\u001b[0m, in \u001b[0;36mget_library_name\u001b[1;34m(organism)\u001b[0m\n\u001b[0;32m    213\u001b[0m             \u001b[38;5;28;01mbreak\u001b[39;00m\n\u001b[0;32m    215\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m database\u001b[38;5;241m.\u001b[39mendswith(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mEnrichr\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[1;32m--> 216\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mLookupError\u001b[39;00m(\n\u001b[0;32m    217\u001b[0m \u001b[38;5;250m        \u001b[39m\u001b[38;5;124;03m\"\"\"No supported database. Please input one of these:\u001b[39;00m\n\u001b[0;32m    218\u001b[0m \u001b[38;5;124;03m                        ('Human', 'Mouse', 'Yeast', 'Fly', 'Fish', 'Worm') \"\"\"\u001b[39;00m\n\u001b[0;32m    219\u001b[0m     )\n\u001b[0;32m    220\u001b[0m \u001b[38;5;66;03m# make a get request to get the gmt names and meta data from Enrichr\u001b[39;00m\n\u001b[0;32m    221\u001b[0m \u001b[38;5;66;03m# old code\u001b[39;00m\n\u001b[0;32m    222\u001b[0m \u001b[38;5;66;03m# response = requests.get('http://amp.pharm.mssm.edu/Enrichr/geneSetLibrary?mode=meta')\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    229\u001b[0m \u001b[38;5;66;03m#     if inst_gmt['isActive'] == True:\u001b[39;00m\n\u001b[0;32m    230\u001b[0m \u001b[38;5;66;03m#         libs.append(inst_gmt['libraryName'])\u001b[39;00m\n\u001b[0;32m    231\u001b[0m lib_url \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m/\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m/datasetStatistics\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m%\u001b[39m (ENRICHR_URL, database)\n",
      "\u001b[1;31mLookupError\u001b[0m: No supported database. Please input one of these:\n                            ('Human', 'Mouse', 'Yeast', 'Fly', 'Fish', 'Worm') "
     ]
    }
   ],
   "source": [
    "gene_sets = gp.get_library_name(\"MsigDb\")\n",
    "\n",
    "gene_set = gene_sets[\"regulation of microtubule depolymerization (GO:0031114)\"]\n",
    "genes = gene_set.genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "ff107493-38a7-46d3-bf27-320e26fbabf8",
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'bio'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[43], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mbio\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m go\n",
      "\u001b[1;31mModuleNotFoundError\u001b[0m: No module named 'bio'"
     ]
    }
   ],
   "source": [
    "from bio import go"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "57fd67bf-5eae-4004-aaf3-158a773ceaae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Collecting bio\n",
      "  Downloading bio-1.7.0-py3-none-any.whl.metadata (5.6 kB)\n",
      "Collecting biopython>=1.80 (from bio)\n",
      "  Downloading biopython-1.83-cp312-cp312-win_amd64.whl.metadata (13 kB)\n",
      "Collecting gprofiler-official (from bio)\n",
      "  Downloading gprofiler_official-1.0.0-py3-none-any.whl.metadata (11 kB)\n",
      "Collecting mygene (from bio)\n",
      "  Downloading mygene-3.2.2-py2.py3-none-any.whl.metadata (10 kB)\n",
      "Requirement already satisfied: pandas in c:\\users\\ashwa\\appdata\\roaming\\python\\python312\\site-packages (from bio) (2.2.0)\n",
      "Collecting pooch (from bio)\n",
      "  Downloading pooch-1.8.1-py3-none-any.whl.metadata (9.5 kB)\n",
      "Requirement already satisfied: requests in c:\\users\\ashwa\\appdata\\roaming\\python\\python312\\site-packages (from bio) (2.31.0)\n",
      "Requirement already satisfied: tqdm in c:\\users\\ashwa\\appdata\\roaming\\python\\python312\\site-packages (from bio) (4.66.1)\n",
      "Requirement already satisfied: numpy in c:\\users\\ashwa\\appdata\\roaming\\python\\python312\\site-packages (from biopython>=1.80->bio) (1.26.4)\n",
      "Collecting biothings-client>=0.2.6 (from mygene->bio)\n",
      "  Downloading biothings_client-0.3.1-py2.py3-none-any.whl.metadata (9.8 kB)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in c:\\users\\ashwa\\appdata\\roaming\\python\\python312\\site-packages (from pandas->bio) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in c:\\users\\ashwa\\appdata\\roaming\\python\\python312\\site-packages (from pandas->bio) (2024.1)\n",
      "Requirement already satisfied: tzdata>=2022.7 in c:\\users\\ashwa\\appdata\\roaming\\python\\python312\\site-packages (from pandas->bio) (2023.4)\n",
      "Requirement already satisfied: platformdirs>=2.5.0 in c:\\users\\ashwa\\appdata\\roaming\\python\\python312\\site-packages (from pooch->bio) (4.2.0)\n",
      "Requirement already satisfied: packaging>=20.0 in c:\\users\\ashwa\\appdata\\roaming\\python\\python312\\site-packages (from pooch->bio) (23.2)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in c:\\users\\ashwa\\appdata\\roaming\\python\\python312\\site-packages (from requests->bio) (3.3.2)\n",
      "Requirement already satisfied: idna<4,>=2.5 in c:\\users\\ashwa\\appdata\\roaming\\python\\python312\\site-packages (from requests->bio) (3.6)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in c:\\users\\ashwa\\appdata\\roaming\\python\\python312\\site-packages (from requests->bio) (2.2.0)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in c:\\users\\ashwa\\appdata\\roaming\\python\\python312\\site-packages (from requests->bio) (2024.2.2)\n",
      "Requirement already satisfied: colorama in c:\\users\\ashwa\\appdata\\roaming\\python\\python312\\site-packages (from tqdm->bio) (0.4.6)\n",
      "Requirement already satisfied: six>=1.5 in c:\\users\\ashwa\\appdata\\roaming\\python\\python312\\site-packages (from python-dateutil>=2.8.2->pandas->bio) (1.16.0)\n",
      "Downloading bio-1.7.0-py3-none-any.whl (279 kB)\n",
      "   ---------------------------------------- 0.0/279.2 kB ? eta -:--:--\n",
      "   -- ------------------------------------- 20.5/279.2 kB ? eta -:--:--\n",
      "   -- ------------------------------------- 20.5/279.2 kB ? eta -:--:--\n",
      "   -- ------------------------------------- 20.5/279.2 kB ? eta -:--:--\n",
      "   -------- ------------------------------ 61.4/279.2 kB 273.8 kB/s eta 0:00:01\n",
      "   ------------- ------------------------ 102.4/279.2 kB 420.8 kB/s eta 0:00:01\n",
      "   ------------------ ------------------- 133.1/279.2 kB 463.7 kB/s eta 0:00:01\n",
      "   ------------------------- ------------ 184.3/279.2 kB 557.9 kB/s eta 0:00:01\n",
      "   --------------------------------- ---- 245.8/279.2 kB 684.7 kB/s eta 0:00:01\n",
      "   --------------------------------- ---- 245.8/279.2 kB 684.7 kB/s eta 0:00:01\n",
      "   -------------------------------------- 279.2/279.2 kB 615.4 kB/s eta 0:00:00\n",
      "Downloading biopython-1.83-cp312-cp312-win_amd64.whl (2.7 MB)\n",
      "   ---------------------------------------- 0.0/2.7 MB ? eta -:--:--\n",
      "    --------------------------------------- 0.0/2.7 MB 1.9 MB/s eta 0:00:02\n",
      "   - -------------------------------------- 0.1/2.7 MB 2.0 MB/s eta 0:00:02\n",
      "   -- ------------------------------------- 0.2/2.7 MB 1.3 MB/s eta 0:00:02\n",
      "   --- ------------------------------------ 0.2/2.7 MB 1.2 MB/s eta 0:00:03\n",
      "   --- ------------------------------------ 0.2/2.7 MB 1.3 MB/s eta 0:00:02\n",
      "   ---- ----------------------------------- 0.3/2.7 MB 1.2 MB/s eta 0:00:03\n",
      "   ---- ----------------------------------- 0.3/2.7 MB 1.1 MB/s eta 0:00:03\n",
      "   ---- ----------------------------------- 0.3/2.7 MB 1.1 MB/s eta 0:00:03\n",
      "   ----- ---------------------------------- 0.4/2.7 MB 983.0 kB/s eta 0:00:03\n",
      "   ------ --------------------------------- 0.4/2.7 MB 994.9 kB/s eta 0:00:03\n",
      "   ------ --------------------------------- 0.5/2.7 MB 960.7 kB/s eta 0:00:03\n",
      "   ------- -------------------------------- 0.5/2.7 MB 971.2 kB/s eta 0:00:03\n",
      "   ------- -------------------------------- 0.5/2.7 MB 943.2 kB/s eta 0:00:03\n",
      "   -------- ------------------------------- 0.6/2.7 MB 906.5 kB/s eta 0:00:03\n",
      "   -------- ------------------------------- 0.6/2.7 MB 904.0 kB/s eta 0:00:03\n",
      "   --------- ------------------------------ 0.6/2.7 MB 922.8 kB/s eta 0:00:03\n",
      "   --------- ------------------------------ 0.7/2.7 MB 877.7 kB/s eta 0:00:03\n",
      "   --------- ------------------------------ 0.7/2.7 MB 877.7 kB/s eta 0:00:03\n",
      "   --------- ------------------------------ 0.7/2.7 MB 802.9 kB/s eta 0:00:03\n",
      "   ---------- ----------------------------- 0.7/2.7 MB 784.1 kB/s eta 0:00:03\n",
      "   ---------- ----------------------------- 0.7/2.7 MB 801.6 kB/s eta 0:00:03\n",
      "   ----------- ---------------------------- 0.8/2.7 MB 824.3 kB/s eta 0:00:03\n",
      "   ------------ --------------------------- 0.8/2.7 MB 816.4 kB/s eta 0:00:03\n",
      "   ------------- -------------------------- 0.9/2.7 MB 828.4 kB/s eta 0:00:03\n",
      "   ------------- -------------------------- 0.9/2.7 MB 833.1 kB/s eta 0:00:03\n",
      "   ------------- -------------------------- 0.9/2.7 MB 828.3 kB/s eta 0:00:03\n",
      "   --------------- ------------------------ 1.0/2.7 MB 850.5 kB/s eta 0:00:02\n",
      "   ---------------- ----------------------- 1.1/2.7 MB 876.5 kB/s eta 0:00:02\n",
      "   ----------------- ---------------------- 1.2/2.7 MB 902.6 kB/s eta 0:00:02\n",
      "   ------------------ --------------------- 1.3/2.7 MB 937.1 kB/s eta 0:00:02\n",
      "   ------------------- -------------------- 1.3/2.7 MB 960.7 kB/s eta 0:00:02\n",
      "   ------------------- -------------------- 1.4/2.7 MB 960.7 kB/s eta 0:00:02\n",
      "   --------------------- ------------------ 1.4/2.7 MB 968.9 kB/s eta 0:00:02\n",
      "   --------------------- ------------------ 1.5/2.7 MB 965.9 kB/s eta 0:00:02\n",
      "   ---------------------- ----------------- 1.5/2.7 MB 979.6 kB/s eta 0:00:02\n",
      "   ----------------------- ---------------- 1.6/2.7 MB 982.9 kB/s eta 0:00:02\n",
      "   ------------------------ --------------- 1.7/2.7 MB 994.9 kB/s eta 0:00:02\n",
      "   ------------------------- -------------- 1.7/2.7 MB 1.0 MB/s eta 0:00:01\n",
      "   --------------------------- ------------ 1.9/2.7 MB 1.0 MB/s eta 0:00:01\n",
      "   --------------------------- ------------ 1.9/2.7 MB 1.0 MB/s eta 0:00:01\n",
      "   ----------------------------- ---------- 2.0/2.7 MB 1.1 MB/s eta 0:00:01\n",
      "   ------------------------------ --------- 2.1/2.7 MB 1.1 MB/s eta 0:00:01\n",
      "   ------------------------------- -------- 2.2/2.7 MB 1.1 MB/s eta 0:00:01\n",
      "   -------------------------------- ------- 2.2/2.7 MB 1.1 MB/s eta 0:00:01\n",
      "   -------------------------------- ------- 2.2/2.7 MB 1.1 MB/s eta 0:00:01\n",
      "   --------------------------------- ------ 2.3/2.7 MB 1.1 MB/s eta 0:00:01\n",
      "   ---------------------------------- ----- 2.3/2.7 MB 1.1 MB/s eta 0:00:01\n",
      "   ---------------------------------- ----- 2.4/2.7 MB 1.1 MB/s eta 0:00:01\n",
      "   ------------------------------------ --- 2.5/2.7 MB 1.1 MB/s eta 0:00:01\n",
      "   ------------------------------------ --- 2.5/2.7 MB 1.1 MB/s eta 0:00:01\n",
      "   ------------------------------------ --- 2.5/2.7 MB 1.1 MB/s eta 0:00:01\n",
      "   ------------------------------------- -- 2.5/2.7 MB 1.1 MB/s eta 0:00:01\n",
      "   ------------------------------------- -- 2.5/2.7 MB 1.1 MB/s eta 0:00:01\n",
      "   ------------------------------------- -- 2.5/2.7 MB 1.0 MB/s eta 0:00:01\n",
      "   ------------------------------------- -- 2.5/2.7 MB 1.0 MB/s eta 0:00:01\n",
      "   ------------------------------------- -- 2.6/2.7 MB 1.0 MB/s eta 0:00:01\n",
      "   ------------------------------------- -- 2.6/2.7 MB 1.0 MB/s eta 0:00:01\n",
      "   -------------------------------------- - 2.6/2.7 MB 983.0 kB/s eta 0:00:01\n",
      "   -------------------------------------- - 2.6/2.7 MB 973.4 kB/s eta 0:00:01\n",
      "   -------------------------------------- - 2.6/2.7 MB 957.0 kB/s eta 0:00:01\n",
      "   ---------------------------------------  2.7/2.7 MB 958.8 kB/s eta 0:00:01\n",
      "   ---------------------------------------  2.7/2.7 MB 946.5 kB/s eta 0:00:01\n",
      "   ---------------------------------------  2.7/2.7 MB 943.3 kB/s eta 0:00:01\n",
      "   ---------------------------------------  2.7/2.7 MB 943.3 kB/s eta 0:00:01\n",
      "   ---------------------------------------  2.7/2.7 MB 943.3 kB/s eta 0:00:01\n",
      "   ---------------------------------------  2.7/2.7 MB 943.3 kB/s eta 0:00:01\n",
      "   ---------------------------------------  2.7/2.7 MB 879.2 kB/s eta 0:00:01\n",
      "   ---------------------------------------- 2.7/2.7 MB 877.6 kB/s eta 0:00:00\n",
      "Downloading gprofiler_official-1.0.0-py3-none-any.whl (9.3 kB)\n",
      "Downloading mygene-3.2.2-py2.py3-none-any.whl (5.4 kB)\n",
      "Downloading pooch-1.8.1-py3-none-any.whl (62 kB)\n",
      "   ---------------------------------------- 0.0/63.0 kB ? eta -:--:--\n",
      "   ------ --------------------------------- 10.2/63.0 kB ? eta -:--:--\n",
      "   -------------------------- ------------- 41.0/63.0 kB 495.5 kB/s eta 0:00:01\n",
      "   ---------------------------------------- 63.0/63.0 kB 483.6 kB/s eta 0:00:00\n",
      "Downloading biothings_client-0.3.1-py2.py3-none-any.whl (29 kB)\n",
      "Installing collected packages: biopython, pooch, gprofiler-official, biothings-client, mygene, bio\n",
      "Successfully installed bio-1.7.0 biopython-1.83 biothings-client-0.3.1 gprofiler-official-1.0.0 mygene-3.2.2 pooch-1.8.1\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install bio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6155553b-95ba-406e-a134-c65459dcf750",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
